alibi_detect/utils/schemas.py

```diff
@@ -0,0 +1,431 @@
+    embedding: Union[str, EmbeddingConfig, None] = None
+    tokenizer: Union[str, TokenizerConfig, None] = None
+    device: Literal['cpu', 'cuda', 'gpu', None] = None
+    preprocess_batch_fn: Optional[str] = None
+    max_len: Optional[int] = None
+    batch_size: Optional[int] = int(1e10)
+    dtype: Optional[str] = None
+
+    # Additional kwargs
+    kwargs: dict = {}
+
+
+class PreprocessConfigResolved(PreprocessConfig):
+    """
+    Resolved schema for preprocess_fn.
+    """
+    src: Callable
+    device: Optional['torch.device'] = None  # Note: `device` resolved for preprocess_drift, but str for detectors
+    model: Optional[SUPPORTED_MODELS] = None  # TODO - Not optional if src is preprocess_drift
+    embedding: Optional[TransformerEmbedding] = None
+    tokenizer: Optional[PreTrainedTokenizerBase] = None
+    preprocess_batch_fn: Optional[Callable] = None
+    dtype: Optional[type] = None
+
+
+class KernelConfig(CustomBaseModel):
+    """
+    Unresolved schema for kernels.
+    """
+    src: str
+
+    # Below kwargs are only passed if kernel == @GaussianRBF
+    sigma: Optional[List[float]] = None
+    trainable: bool = False
+
+    # Additional kwargs
+    kwargs: dict = {}
+
+
+class KernelConfigResolved(KernelConfig):
+    """
+    Resolved schema for kernels.
+    """
+    src: Callable
+    sigma: Optional[np.ndarray] = None
+
+
+class DeepKernelConfig(CustomBaseModel):
+    """
+    Unresolved schema for DeepKernels (see utils.[backend].kernels.DeepKernel).
+    """
+    proj: Union[str, ModelConfig]
+    kernel_a: Union[str, KernelConfig] = "@utils.tensorflow.kernels.GaussianRBF"
+    kernel_b: Union[str, KernelConfig, None] = None
+    eps: Union[float, str] = 'trainable'
+
+
+class DeepKernelConfigResolved(DeepKernelConfig):
+    """
+    Resolved schema for DeepKernels (see utils.[backend].kernels.DeepKernel).
+    """
+    proj: SUPPORTED_MODELS
+    kernel_a: Union[Callable, KernelConfigResolved]
+    kernel_b: Union[Callable, KernelConfigResolved, None] = None
+
+
+class DriftDetectorConfig(DetectorConfig):
+    """
+    Unresolved base schema for drift detectors.
+    """
+    # args/kwargs shared by all drift detectors
+    x_ref: str
+    p_val: float = .05
+    x_ref_preprocessed: bool = False
+    preprocess_fn: Union[str, PreprocessConfig, None]
+    input_shape: Optional[tuple] = None
+    data_type: Optional[str] = None
+
+
+class DriftDetectorConfigResolved(DriftDetectorConfig):
+    """
+    Resolved base schema for drift detectors.
+    """
+    x_ref: Union[np.ndarray, list]
+    preprocess_fn: Union[Callable, PreprocessConfigResolved, None]
+
+
+class KSDriftConfig(DriftDetectorConfig):
+    """
+    Unresolved schema for KSDrift detector.
+    """
+    preprocess_at_init: bool = True
+    update_x_ref: Optional[Dict[str, int]] = None
+    correction: str = 'bonferroni'
+    alternative: str = 'two-sided'
```

Comment:
Should be `Literal`s?

Reply:
Let's open a new issue to update these to `Literal`s.

Reply:
https://github.com/SeldonIO/alibi-detect/issues/463